File: Ksenia Kuzmina/Term 4/Task 6 (4.1)/PeerToPeerChat/Network/Connection.cs
using System;
using System.Collections.Generic;
using System.Linq;
using System.Text;
using System.Threading.Tasks;
using System.Net.Sockets;
using System.Net;
using System.Reflection;
using System.Net.WebSockets;

namespace PeerToPeerChat.Network
{
    public class Connection
    {
        private Socket socket;

        public Connection(IPEndPoint address)
        {
            socket = new Socket(address.AddressFamily, SocketType.Stream, ProtocolType.Tcp);
            socket.Connect(address);
        }

        public Connection(Socket socket)
        {
            this.socket = socket;
        }

        public void Send(string data)
        {
            try
            {
                Console.WriteLine("Sending: " + data);

                byte[] messageSent = Encoding.UTF8.GetBytes(data);
                int byteSent = socket.Send(messageSent);

                if (byteSent < 0) throw new Exception("Remote socket closed");
            }
            catch (Exception e) 
            {
                Console.WriteLine(e.Message);
            }
        }

        public string Receive()
        {
            var messageReceived = new byte[1024];
            var byteRecv = socket.Receive(messageReceived);
            var res = Encoding.UTF8.GetString(messageReceived, 0, byteRecv);

            if (byteRecv == 0)
	            throw new Exception("Remote socket closed");
            Console.WriteLine("Received " + res);

            return res;
        }

        public IPEndPoint LocalAddress()
        {
            return socket.LocalEndPoint as IPEndPoint;
        }

        public IPEndPoint RemoteAddress()
        {
            return socket.RemoteEndPoint as IPEndPoint;
        }

        public void Close()
        {
            socket.Close();
        }
    }
}

File: IgorKorobitsyn/Block4/Task6/src/main/kotlin/peerToPeerChat/Connection.kt
package peerToPeerChat
import java.net.InetSocketAddress
import java.net.Socket
import java.nio.charset.Charset

class Connection(
    val socket: Socket
) {

    constructor(address: InetSocketAddress) : this(Socket(address.address, address.port)) {
        //this.socket.close()
    }


    fun send(data: String) {
        try {
            println("Sending: $data")

            val messageSent = data.toByteArray(Charset.forName("UTF-8"))
            socket.getOutputStream().write(messageSent)

        } catch (e: Exception) {
            println(e.message)
            println("In Connection.send")
        }
    }

    fun receive(): String {
        val messageReceived = ByteArray(1024)
        val byteRecv = socket.getInputStream().read(messageReceived)
        if (byteRecv == -1) throw Exception("Remote socket closed") //return ""
        //if (byteRecv == 0) throw Exception("string is empty")
        val res = String(messageReceived, 0, byteRecv, Charset.forName("UTF-8"))


        println("Received $res")

        return res
    }

    fun localAddress(): InetSocketAddress {
        return socket.getLocalSocketAddress() as InetSocketAddress
    }

    fun remoteAddress(): InetSocketAddress {
        return socket.getRemoteSocketAddress() as InetSocketAddress
    }

    fun close() {
        socket.close()
    }
}
